service/sign/handlers_sign.go

```diff
@@ -133,15 +177,102 @@ func Serial(w http.ResponseWriter, r *http.Request) response.ErrorResponse {
+		}
+
+		// Double check the serial assertion
+		serialAssert, ok := assertions["serial"]
+		if !ok {
+			const msg = "The current serial assertion can't be empty for a remodeling request"
+			svlog.Message("SIGN", "invalid-assertion", msg)
+			return response.ErrorResponse{Success: false, Code: response.ErrorInvalidAssertion.Code, Message: msg, StatusCode: http.StatusBadRequest}
+		}
+
+		// Validate the original model by checking that it exists on the database
+		originalModelAssert, errResponse := findModel(originalBrandID, originalModel, originalSerial, apiKey)
+		if !errResponse.Success {
+			svlog.Message("SIGN", "invalid-assertion", "original model is not valid")
+			return errResponse
+		}
+
+		// Validate the new model: it must be defind in the sub-store of the orignal model
+		substore, err := datastore.Environ.DB.GetSubstore(originalModelAssert.ID, originalSerial)
+		if err != nil {
+			svlog.Message("PIVOT", "invalid-substore", "Cannot find sub-store mapping for the model")
+			return response.ErrorInvalidSubstore
+		}
+
+		// Check if find model maches requested model
+		if serialReq.HeaderString("model") != substore.ModelName {
+			const msg = "Requested model is invalid"
+			svlog.Message("SIGN", "invalid-assertion", msg)
+			return response.ErrorResponse{Success: false, Code: response.ErrorInvalidAssertion.Code, Message: msg, StatusCode: http.StatusBadRequest}
+		}
+
+		// Check original-* fields matchine old serial
+		if serialAssert.HeaderString("model") != originalModel {
+			const msg = "Original model is invalid"
+			svlog.Message("SIGN", "invalid-assertion", msg)
+			return response.ErrorResponse{Success: false, Code: response.ErrorInvalidAssertion.Code, Message: msg, StatusCode: http.StatusBadRequest}
+		}
+		if serialAssert.HeaderString("serial") != originalSerial {
+			const msg = "Original serial number is invalid"
+			svlog.Message("SIGN", "invalid-assertion", msg)
+			return response.ErrorResponse{Success: false, Code: response.ErrorInvalidAssertion.Code, Message: msg, StatusCode: http.StatusBadRequest}
+		}
+		if serialAssert.HeaderString("brand-id") != originalBrandID {
+			const msg = "Original brand-id is invalid"
+			svlog.Message("SIGN", "invalid-assertion", msg)
+			return response.ErrorResponse{Success: false, Code: response.ErrorInvalidAssertion.Code, Message: msg, StatusCode: http.StatusBadRequest}
+		}
+		// Check that the device key is the same between serial-request and old serial
+		if serialAssert.HeaderString("device-key") != serialReq.HeaderString("device-key") {
+			const msg = "Device-key is invalid"
+			svlog.Message("SIGN", "invalid-assertion", msg)
+			return response.ErrorResponse{Success: false, Code: response.ErrorInvalidAssertion.Code, Message: msg, StatusCode: http.StatusBadRequest}
+		}
+
+		// Sign the sended old serial assertion and compare the signature with sended signature
```

Comment:
it should be possible to check this with something like:

err := asserts.SignatureCheck(serialAssert,  ...KeypairDB.PublicKey(KEY_ID))